### PR TITLE
acme-dns: init at 0.8

### PR DESCRIPTION
**Pull request by emilazy**
###### Motivation for this change

[acme-dns](https://github.com/joohoi/acme-dns) allows for easy and secure configuration of ACME DNS-01 validation, without the need for DNS-provider-specific hooks or API keys with total control over DNS zones. These commits add acme-dns client integration to the `security.acme` module, a package and NixOS module for the server, and a test that ties them all together.

There are a few things worth noting about this module compared to the average NixOS module:

* Uses the same option names and structure as the upstream project; this reduces the amount of boilerplate and templating required in the module and allows for easier porting of configurations to NixOS, at the expense of option names that are less harmonized with the rest of NixOS.

* Hardened with an extensive set of sandboxing options, under the general principle that a publicly-exposed network service that essentially has complete power over your certificate renewal ought to run in as restricted an environment as possible. Unfortunately, at present this mostly means it's better-hardened than the web servers it helps issue keys for...

There's also some miscellaneous refactoring and clean-ups to the acme test included. I'm successfully using this to issue certificates with DNS-01 validation on my server.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by emilazy** [comment]
@GrahamcOfBorg test acme

**Comment by emilazy** [comment]
@GrahamcOfBorg test acme acme-dns

**Comment by emilazy** [comment]
Bugs found during week writing PR: 50
Bugs found during 5 minutes after submitting PR: also 50

@GrahamcOfBorg test acme acme-dns

**Comment by emilazy** [comment]
Rebased to fix conflicts.

@GrahamcOfBorg test acme acme-dns

**Comment by yegortimoshenko** [comment]
Pign @Ma27 @mweinelt :)

**Comment by Ma27** [comment]
I'll take a look today :)

**Comment by emilazy** [comment]
Rebased.

@GrahamcOfBorg test acme acme-dns

cc @Ma72 @mweinelt @arianvp @immae @m1cr0man @aanderse @flokli (perhaps we should have a maintainer group like GNOME and corresponding @NixOS/acme-maintainers ping?); it's a bit fiddly to keep the test cleanups and module changes updated against churn, so I'd like to get this upstreamed sooner rather than later if possible :) I can split the non-acme-dns changes into a separate PR if desired (otherwise when this lands I'll backport them to 20.03).

**Comment by emilazy** [comment]
@GrahamcOfBorg test acme acme-dns

(#85376)

**Comment by aanderse** [comment]
@emilazy contrary to your reasoning I prefer we just put this on the back burner until the `acme` module goes through the cleanup we had suggested. I would prefer the `acme` to not have any changes until `20.03` is out and goes through a small stabilization period. If you have some serious pressing need to get this in `master` we should discuss that, but the `acme` module has enough "going on" with respect to the release which is likely coming in the next few days.

I'll mention I carefully chose the word "prefer" over "insist" above. Please don't feel I'm insisting. I'm just stating my opinion, and hopefully a reasonable explanation of why.

Your idea about a maintainers team is a fantastic idea and I fully support that.

**Comment by emilazy** [comment]
My feeling was that the test cleanups are pretty simple and it'll be a pain to have to backport changes across them if they're not included in 20.03. I do understand the reticence to touch things so close to release, but feel a bit like the recent churn (#84781, #85185, #85333, #85366, ...) is kinda making that a moot point anyway :)

I'll defer to what other people think as far as 20.03 goes, but it'd at least be nice to get this into master (we'll have to deal with the test discrepancies if it's merged at any time in the next release cycle anyway).

**Comment by immae** [comment]
I’m only a passing by (not even a real contributor yet :p ) who happened to get poked, so I’ll just give my feeling of it:
- like @aanderse I agree it’s a little dangerous to add such a big work to a release branch that is in beta (so it should by definition only get bugfixes, which all the PR’s you point to are actually @emilazy  :p )
- However you did a very big work, and the acme module in its current state is brand new: it will break many setups (I reported a new case just today for instance) even though we’re trying to limit the effect of the migration. As far as I can see your work will only add new features and no change of the existing, so it won’t break anything new (only things which doesn’t exist yet).
- The fact that the acme module is a critical one and is a bit instable makes me think that we should try as long as possible to remain even between master and release-20.03, to make subsequent fixes easier to write (it’s already not the case but it could be fixed), since they *will* come for sure

Considering that, I think it better to either merge it now or in a few weeks/months after the dust of the ACME change settles (preference for now, since it will become cumbersome to follow every change), and in any case make both the pre-release and master identical as far as the acme module is concerned.

**Comment by yegortimoshenko** [comment]
I'd prefer to have this merged in this release cycle, given there are more risky PRs that are being merged in 20.03 that touch ACME (this one only features new codepath and test renames, so low-risk).

File: nixos/tests/common/acme/server/default.nix
Comment (arianvp) [inline nixos/tests/common/acme/server/default.nix:3]: This is not true anymore since `pebble`. `pebble` generates its own CA on startup; and we download that from their internal endpoint.  This testing code is a bit messy becuase of this.

Comment (arianvp) [reply nixos/tests/common/acme/server/default.nix:3]: _edit_: nvm I see we configure pebble to use this cert

File: nixos/tests/acme.nix
Comment (arianvp) [inline nixos/tests/acme.nix:207]: Given we now have 

```

  security.pki.certificateFiles = [ acme-ca ];
}
```

shouldn't we be able to remove these `--cacert /tmp/ca.crt` calls? If not, why not?

File: nixos/tests/common/acme/server/default.nix
Comment (emilazy) [reply nixos/tests/common/acme/server/default.nix:3]: My understanding is that this is the CA that issues the static TLS certificate that is used to authenticate access to `https://acme.test/` itself, rather than the CA that Pebble uses to issue ACME certificates. Hope this makes sense... the distinction confused me at first too.

File: nixos/tests/acme.nix
Comment (emilazy) [reply nixos/tests/acme.nix:207]: For the same reason as my previous reply :)

File: nixos/tests/common/acme/server/default.nix
Comment (emilazy) [reply nixos/tests/common/acme/server/default.nix:3]: See https://github.com/letsencrypt/pebble#avoiding-client-https-errors; we just generate our own CA rather than using the bundled one. That might not be worth it, though.

File: pkgs/tools/admin/pebble/default.nix
Comment (arianvp) [inline pkgs/tools/admin/pebble/default.nix:19]: Why change to `buildGoModule` if `buildGoPackage` with the vendored dependencies still work?   Also see https://github.com/NixOS/nixpkgs/issues/84826

Comment (arianvp) [reply pkgs/tools/admin/pebble/default.nix:19]: I'm not too opinionated about it; but other people seem to be very against introducing any new `buildGoModule` things into `nixpkgs`

Comment (emilazy) [reply pkgs/tools/admin/pebble/default.nix:19]: I tend to agree with the linked issue and dislike fixed-output derivations, was just following the manual's recommendation for best practices. Happy to change this back to `buildGoPackage`.

File: nixos/tests/common/acme/server/default.nix
Comment (emilazy) [reply nixos/tests/common/acme/server/default.nix:3]: Okay, we can't use the upstream certificates because they're only valid for `pebble` and `localhost`. We could reuse their CA since they include the private key and just issue our own certificates for the API, but at that point the simplification is marginal and I'm going to say it's out of scope for this PR.

File: pkgs/tools/admin/pebble/default.nix
Comment (emilazy) [reply pkgs/tools/admin/pebble/default.nix:19]: I remember why now: `buildGoPackage` doesn't support the official Go module dependency infrastructure at all, which seems unfortunate, although not a blocker in this case. What confuses me is that there was no vendored `deps.nix` even before 4e408d9d11d71fd8e0a15121bda7d58f1bdcc507...

Comment (emilazy) [reply pkgs/tools/admin/pebble/default.nix]: Now uses `buildGoPackage` + vgo2nix.

@GrahamcOfBorg test acme acme-dns

**Comment by emilazy** [comment]
@GrahamcOfBorg test acme acme-dns

**Comment by m1cr0man** [comment]
Ah, I was about to ask how this plays with existing DNS servers on a host, but now I see it's registered as a dnsProvider :)

File: pkgs/tools/admin/pebble/default.nix
Comment (arianvp) [reply pkgs/tools/admin/pebble/default.nix]: I think it's because they have everything in a `vendor` folder

Comment (emilazy) [reply pkgs/tools/admin/pebble/default.nix]: Oh, huh... Go is so weird. Dose this mean that the `deps.nix` is incorrect?

**Comment by emilazy** [comment]
@m1cr0man For the client integration you have to manually specify `ACME_DNS_API_BASE` in the `credentialsFile`, yeah. On the server end, you do have to make sure to avoid clashing with other local DNS servers, for what it's worth, which can be tricky since you can't use any port other than 53 for either NS records or `/etc/resolv.conf` nameservers. Here's what I do to use a local `dnscrypt-proxy2` while still serving `acme-dns` on port 53:

```nix
{
  networking.nameservers = [ "::1" ];

  services.dnscrypt-proxy2 = {
    enable = true;
    settings = {
      # We use port 51 to avoid conflicting with patchouli's acme-dns,
      # and forward 53 to it for local connections with iptables rules.
      listen_addresses = [ "[::1]:51" ];
      # ...
    };
  };

  # Forward loopback traffic on port 53 to dnscrypt-proxy2.
  networking.firewall.extraCommands = ''
    ip6tables --table nat --flush OUTPUT
    ${lib.flip (lib.concatMapStringsSep "\n") [ "udp" "tcp" ] (proto: ''
      ip6tables --table nat --append OUTPUT \
        --protocol ${proto} --destination ::1 --destination-port 53 \
        --jump REDIRECT --to-ports 51
    '')}
  '';
}

**Comment by m1cr0man** [comment]
That's a really interesting solution. Might be worth throwing that in a wiki post? I doubt I'll be the only one to consider running acme-dns and a DNS server on the same host

**Comment by emilazy** [comment]
Note that this workaround is only useful in the case where you're running a local DNS proxy or recursive resolver; if you're running an authoritative DNS server that you want to be accessible from other hosts, then this solution doesn't work, since there's a fundamental clash on external port 53; your only options then are to have them listen on different IP addresses, or running the acme-dns server on loopback and delegating/forwarding a zone to it (like the acme-dns test does with CoreDNS).

That said, I agree it'd be nice to have it documented somewhere; I was hoping it could be integrated into the NixOS `dnscrypt-proxy2` module somehow to reliably set up local DNS regardless of what you're running on port 53, but ran into issues I don't quite remember when trying to universalize the setup to the extent that it could be upstreamed. Looking for somewhere to document it on the wiki, the [DNSCrypt](https://nixos.wiki/wiki/DNSCrypt) article uses the old, unmaintained version of dnscrypt-proxy, and the [DNS over TLS](https://nixos.wiki/wiki/DNS_over_TLS) article recommends the [security tire fire](https://twitter.com/FiloSottile/status/1235044425509810176) that is dnsmasq... looks like I have an article to write :)

File: pkgs/tools/admin/pebble/default.nix
Comment (nlewo) [reply pkgs/tools/admin/pebble/default.nix]: Because they could be afraid a dependency out of their control disappear from GitHub;)
So, in this case, I think `goBuildPackage` without any `deps.nix` file should do the job.

**Comment by emilazy** [comment]
Ditched the pebble `deps.nix` per https://github.com/NixOS/nixpkgs/pull/83474#discussion_r410501948. I really hope I don't have to rebase this too many more times... ^^;

@GrahamcOfBorg test acme acme-dns

**Comment by flokli** [comment]
I agree with @aanderse on https://github.com/NixOS/nixpkgs/pull/83474#issuecomment-614844730.

TBH, the ACME module has received quite some last minute fixes to unbreak some bugs in 20.03, we still haven't ruled out all the issues mentioned in #84633.

IMHO, we should do the clean up and refactoring needed in the test and module (carefully backporting if possible - I don't see we couldn't do that at least for the tests) - and then carefully start introducing more features making use of the cleaner interface.

Even though I find `acme-dns` very appealing and am looking forward to use it in the future, I'm afraid of introducing more breakages by accident, and we should really let the dust settle a bit and collect feedback by using it in unstable, instead of rushing to backport things into 20.03.

**Comment by emilazy** [comment]
To be clear, the parts I'm most interested in backporting are the non-acme-dns commits which simply clean up and improve the test; otherwise, any changes to the test will have to be backported across a bunch of conflicting changes for the whole 20.03 release cycle.

**Comment by flokli** [comment]
@emilazy yes, I'm all for improving the tests, and see no reason for not backporting this into 20.03.

Can you move all these commits into a separate PR, and rebase this one on top of it?
They're currently a bit interleaved with the acme-dns addition, and having this two separate PRs makes reviewing things a lot easier…

**Comment by emilazy** [comment]
FWIW, they're not interleaved in the logical git order:

```
emily@renko ~/s/nixpkgs (add-acme-dns)> git log --oneline
b4ffcb559f1 (HEAD -> add-acme-dns, origin/add-acme-dns) nixos/acme-dns: init
56cf244e319 acme-dns: init at 0.8
7cc9da0af6b nixos/acme: add acme-dns client integration
21f183a3fe4 nixos/tests/common/acme: don't set nameservers for client
e6d5e83cf10 nixos/tests/common/acme: enable Pebble strict mode
695fd78ac45 nixos/tests/acme: use CAP_NET_BIND_SERVICE
d0f04c1623a nixos/tests/acme: use *.test domains
352e30df8a3 nixos/tests/acme: don't restrict to x86_64
6285d5eabda pebble: v2.2.2 -> v2.3.0
```

GitHub is just really bad at handling rebases, since it always sorts by the commit date. I'll try to open a separate PR today, though.

**Comment by flokli** [comment]
Urgh, then the "Commits" tab makes no sense currently. Thanks for opening a separate PR!

**Comment by emilazy** [comment]
Rebased now that #85503 has been merged; diff should hopefully be much less intimidating now. It still touches the ACME module, so I can understand holding off until it's been reworked, though it shouldn't change behaviour at all for anyone not already using DNS-01 validation and mostly just introduces a new code path for acme-dns.

@GrahamcOfBorg test acme acme-dns

File: nixos/modules/security/acme.nix
Comment (flokli) [inline nixos/modules/security/acme.nix:443]: Can the registration logic moved to the lego binary or some helper binary installed if with `acme-dns`?

I'm a bit afraid of the more flaky parts here to break in funny ways, and assume more people would benefit from being able to check the DNS configuration w.r.t. acme-dns.

File: nixos/modules/services/security/acme-dns.nix
Comment (flokli) [inline nixos/modules/services/security/acme-dns.nix:372]: We can probably drop these comments.

File: nixos/modules/security/acme.nix
Comment (emilazy) [reply nixos/modules/security/acme.nix:443]: lego's [goacmedns](https://github.com/cpu/goacmedns) dependency isn't packaged separately, but if it was then we could use `goacmedns-register(1)`. It would only affect 4 lines here, so I figured it wasn't worth it as the API is just an empty POST. If you think it'd be better then I can package and use it.

Unfortunately lego handles this *really* badly natively; it doesn't even bother registering until renew time (so you have to scramble to set up CNAME records at unpredictable intervals on initial migration to acme-dns certificates) and doesn't check the CNAME records at all (instead, it just prints the record unconditionally on initial registration and then fails mysteriously if you don't set it up). It would be nice if we could delegate some of this to lego, but given that we're already having to architect our own solution to forcing renewals on configuration changes, etc., I think some of this complexity is unavoidable to ensure a reasonable user experience unless lego changes its interface/operating model.

Comment (emilazy) [reply nixos/modules/security/acme.nix:443]: Opened https://github.com/go-acme/lego/issues/1119, https://github.com/go-acme/lego/issues/1120.

File: nixos/modules/services/security/acme-dns.nix
Comment (emilazy) [reply nixos/modules/services/security/acme-dns.nix:372]: The intent here is to allow cross-referencing against the systemd documentation so that it's easier to check if any hardening option is missing (e.g. because of options being added to new systemd versions), and to ensure that options aren't incorrectly dropped when other sandboxing options that imply them have to be adjusted. I copied this pattern from another local WIP module's sandboxing options and found them helpful for ensuring the service hardening options covered everything they can and made sense together, but I can drop them if you think they're too noisy.

**Comment by emilazy** [comment]
Rebased with some tweaks to the hardening options; `systemd-analyze security acme-dns.service` before:

> → Overall exposure level for acme-dns.service: 2.3 OK 🙂

and after:

> → Overall exposure level for acme-dns.service: 1.6 OK 🙂

This is still technically 0.2 higher than it should be, since `ProtectClock` is redundant with the other hardening settings, but even explicitly setting it didn't fix it. I'm not sure why (maybe our systemd disables `ProtectClock` somehow, or maybe there's some clash with other options?), but I filed https://github.com/systemd/systemd/issues/15758.

I also tried moving over to `buildGoPackage`/`vgo2nix` per https://github.com/NixOS/nixpkgs/issues/84826, but ran into https://github.com/nix-community/vgo2nix/issues/39.

cc @NixOS/acme; how do we want to handle this PR? Most of it is entirely independent from the ACME client implementation, the parts that aren't largely consist of changing behaviour in acme-dns-specific code paths, and even those parts are independent enough that they would remain largely the same after any rewrite of the client module. I can understand not having much appetite for increased functionality and complexity in the module until it's in better shape, but I'm unclear on how actively that's being worked on. I'd certainly be happy to update the client integration here to work better with a reworked client module, but it feels unfortunate to block improvements that shouldn't change the behaviour for common configurations if nobody's actively working on a concrete refactor.

Right now we recommend a BIND setup in the manual that is probably a lot less secure than this and certainly a lot more fuss. It'd be possible to land the server module without any of the client integration and let people set it up themselves with lego, but as I explained in https://github.com/NixOS/nixpkgs/pull/83474#discussion_r411509158, the user experience of this is so bad that I find it hard to recommend. I'm totally fine if we decide it's best to wait to integrate this, just want to make sure we come to a concrete decision and that we're not all collectively waiting on each other.

@GrahamcOfBorg test acme acme-dns

**Comment by emilazy** [comment]
A note: it turns out that `DynamicUser=true` silently overrides the `TemporaryFileSystem=/:ro` and makes some of the filesystem sandboxing inert. Very frustrating behaviour on systemd's part, but I'm working to address it and also use stricter confinement.

**Comment by arianvp** [comment]
I'm actually OK with merging this btw. It seems pretty standalone (eventhough it makes changes to the `acme.nix` file) and  doesn't change anything in the module fundamentally

**Comment by flokli** [comment]
Yeah, also no problem with merging this, if it doesn't complicate the acme module refactor.

@emilazy, could you add some documentation on how to use this to `nixos/modules/security/acme.xml`? Since #86347 we have some documentation on the DNS validation, with `acme-dns` added, we should document it as well.

**Comment by emilazy** [comment]
Will add some docs once I get the sandboxing options working properly again.

**Comment by flokli** [comment]
@emilazy do you mean, this should wait until https://github.com/NixOS/nixpkgs/pull/87661 has been merged?

This needs a rebase btw.

**Comment by emilazy** [comment]
Okay, this should at long last be ready for review again. I switched away from `DynamicUser`, integrated `systemd-confinement`, and did more thorough testing of the sandboxing options. Ironically, the `systemd-analyze security` exposure level went up slightly due to no longer being able to set `ProtectSystem`, but the practical service isolation has gone up.

I also added some documentation to the ACME section of the manual; I replaced the BIND section with it as suggested in #86347, since I think it'll probably be more appropriate for most users than running a full authoritative DNS server for their domains, and the manual is already quite long. I'd appreciate feedback on this, since it was written quite quickly and there's a fair amount of arcane detail to cover in a small amount of space. In particular, I'm not confident the advice for DNS providers that don't fully support glue records (as e.g. Cloudflare doesn't) is clear enough.

> @emilazy do you mean, this should wait until #87661 has been merged?

Not sure that PR will land as-is, at least not imminently, and if it does this won't be the only module that could be refactored to use it, so it should be fine to merge as-is.

@ofborg test acme acme-dns

**Comment by emilazy** [comment]
Fixed a dangling reference to `acmeDnsRegisterService` in a comment and changed the description of the `acme-dns-*` services to `"Ensure acme-dns Credentials for ${cert}"` (as opposed to "Register ..."), since it's really doing "check if present and register if not"; ideas for better descriptions welcome, this is just the first thing I could think of that wasn't misleading.

File: nixos/modules/security/acme.nix
Comment (flokli) [reply nixos/modules/security/acme.nix:443]: Please add comments to these issues into the comment at the beginning of the block.

File: nixos/modules/security/acme.nix
Comment (flokli) [inline nixos/modules/security/acme.nix:337]: This dependency shouldn't be added if acme-dns isn't running on the same machine.

**Review by flokli — CHANGES_REQUESTED** [review]

**Comment by flokli** [comment]
> I also added some documentation to the ACME section of the manual; I replaced the BIND section with it as suggested in #86347, since I think it'll probably be more appropriate for most users than running a full authoritative DNS server for their domains, and the manual is already quite long.

I don't think it would hurt to keep both. I propose to document `acme-dns` first, but keep the BIND docs in a separate section.

**Comment by m1cr0man** [comment]
> > I also added some documentation to the ACME section of the manual; I replaced the BIND section with it as suggested in #86347, since I think it'll probably be more appropriate for most users than running a full authoritative DNS server for their domains, and the manual is already quite long.
> 
> I don't think it would hurt to keep both. I propose to document `acme-dns` first, but keep the BIND docs in a separate section.

Another option is to move it to a wiki page for SSL certs. Considering the importance/size of this module as a whole it would be nice to have regardless.

Also I would argue that there are certainly people going in corporate or self hosted environments that do have existing DNS infrastructure which they would want their challenges to go through. You could definitely correlate this to needing wildcard certs in the first place.

File: nixos/modules/services/security/acme-dns.nix
Comment (dminuoso) [inline nixos/modules/services/security/acme-dns.nix:182]: This should be called `listen` or `listenAddress` instead.

File: pkgs/servers/dns/acme-dns/default.nix
Comment (SuperSandro2000) [inline pkgs/servers/dns/acme-dns/default.nix:22]: Setting homepage explicitly would allow me to visit it right now without copy pasting the url togeter.

File: pkgs/servers/dns/acme-dns/default.nix
Comment (SuperSandro2000) [inline pkgs/servers/dns/acme-dns/default.nix:25]: This is not required to be set and actually wrong. Go is only available on linux and darwin and duplicating that is not required.

**Comment by SuperSandro2000** [comment]
stale and with merge conflicts, going to close it

**Comment by emilazy** [comment]
I stopped using acme-dns quite a long while ago; I'd be happy for someone else to take over this code if they find it useful but I'm unlikely to put more work into the PR myself. Sorry for letting it rot without conclusion.

**Comment by emilazy** [comment]
It should be possible to use a regular DNS server for the ACME zone (e.g. with lego's RFC 2136 support); I believe some of the "big cloud" DNS providers (e.g. I believe Azure can do this, and probably Amazon and Google too?) let you set up fine-grained enough API key permissions to use with a CNAME'd zone, too. I never got around to setting it up - I just fell back to HTTP validation once this PR bitrotted enough.

I forget exactly why I decided against pursuing acme-dns further, but I think it was due to the project's dormancy (at the time; it has had activity since, although it still doesn't seem to be very actively developed and the maintainer was apparently planning a complete rewrite last year) and some lacking features and/or infelicities in its DNS implementation. I also planned to move away from lego in general for ACME so I didn't really have much use left for this PR.

**Comment by m1cr0man** [comment]
FWIW, the acme test suite uses rfc2136 for DNS-01 validation and [the manual covers](https://nixos.org/manual/nixos/stable/index.html#module-security-acme-config-dns) configuring it too. Potentially you could set up an rfc2136-capable DNS server within NixOS to do what acme-dns achieves, and in the same fashion (CNAME delegation).

If there was still interest in an acme-dns-like solution, we could just write a manual section on how to configure a minimal bind to do the job.

**Comment by dminuoso** [comment]
Granular API permissions on DNS is virtually non-existent amongst most providers.

While we did work on acme-dns for internal use for a while, we have long switched to PowerDNS with a custom granular authorization proxy written in Haskell that wraps the PowerDNS API: https://github.com/wobcom/powerdns-gerd

This allows us to grant individual users (or machines in our case) very limited access to the PowerDNS API (constrained by endpoint and a domain name pattern possibly with wildcards)

**Comment by emilylange** [comment]
I redid parts of this in #234207 in case anyone is still interested :)

Without lego/`security.acme` though.

**Pull request closed without merging**
